components/renku_data_services/session/k8s_client.py

```diff
@@ -0,0 +1,268 @@
+    scalable: bool = False
+    endpoint: str = "buildruns"
+
+
+class _ShipwrightClientBase:
+    """Client for managing ShipWright resources in kubernetes."""
+
+    def __init__(self, namespace: str):
+        self.namespace = namespace
+        self.sanitize = ApiClient().sanitize_for_serialization
+
+    async def create_build_run(self, manifest: BuildRun) -> BuildRun:
+        """Create a new ShipWright BuildRun."""
+        manifest.metadata.namespace = self.namespace
+        build_run = await ShipwrightBuildRunV1Beta1Kr8s(manifest.model_dump(exclude_none=True, mode="json"))
+        build_run_name = manifest.metadata.name
+        try:
+            await build_run.create()
+        except ServerError as e:
+            logging.exception(f"Cannot create the image build {build_run_name} because of {e}")
+            raise CannotStartBuildError(message=f"Cannot create the image build {build_run_name}")
+        await build_run.refresh()
+        build_resource = await retry_with_exponential_backoff_async(lambda x: x is None)(self.get_build_run)(
+            build_run_name
+        )
+        if build_resource is None:
+            raise CannotStartBuildError(message=f"Cannot create the image build {build_run_name}")
+        return build_resource
+
+    async def get_build_run(self, name: str) -> BuildRun | None:
+        """Get a ShipWright BuildRun."""
+        try:
+            build = await ShipwrightBuildRunV1Beta1Kr8s.get(name=name, namespace=self.namespace)
+        except NotFoundError:
+            return None
+        except ServerError as e:
+            if not e.response or e.response.status_code not in [400, 404]:
+                logging.exception(f"Cannot get the build {name} because of {e}")
+                raise IntermittentError(f"Cannot get build {name} from the k8s API.")
+            return None
+        return BuildRun.model_validate(build.to_dict())
+
+    async def list_build_runs(self, label_selector: str | None = None) -> list[BuildRun]:
+        """Get a list of ShipWright BuildRuns."""
+        try:
+            builds = await ShipwrightBuildRunV1Beta1Kr8s.list(namespace=self.namespace, label_selector=label_selector)
+        except ServerError as e:
+            if not e.response or e.response.status_code not in [400, 404]:
+                logging.exception(f"Cannot list builds because of {e}")
+                raise IntermittentError("Cannot list builds")
+            return []
+        output = [BuildRun.model_validate(b.to_dict()) for b in builds]
+        return output
+
+    async def delete_build_run(self, name: str) -> None:
+        """Delete a ShipWright BuildRun."""
+        build = await ShipwrightBuildRunV1Beta1Kr8s(dict(metadata=dict(name=name, namespace=self.namespace)))
+        try:
+            await build.delete(propagation_policy="Foreground")
+        except ServerError as e:
+            logging.exception(f"Cannot delete build {name} because of {e}")
+            raise DeleteBuildError()
```

Comment:
suggestion: we should handle the not found case here and not fail - just return None. The thing we want to remove is not there so we are good.